src/mem/arena.rs

```diff
@@ -66,15 +110,80 @@ pub trait Arena {
     /// # let mut arena = BumpArena::new(&mut data);
     /// // If the first alloc succeeds (regardless of the value of `len`), then
     /// // the subsequent alloc after resetting the arena must also succeed.
-    /// assert!(arena.alloc(64).is_ok());
+    /// assert!(arena.alloc_aligned(64, 1).is_ok());
     /// arena.reset();
-    /// assert!(arena.alloc(64).is_ok());
+    /// assert!(arena.alloc_aligned(64, 1).is_ok());
     /// ```
     fn reset(&mut self);
 }
 
 assert_obj_safe!(Arena);
 
+/// Convenience functions for arenas, exposed as a trait.
+///
+/// Note that this trait is implemened for `&impl Arena`, which is the reason
+/// for the slightly odd signature.
+pub trait ArenaExt<'arena> {
```

Comment:
Instead of using an extension trait, you can put these all into Arena directly as provided methods:

```rust
pub trait Arena {
    fn alloc_aligned(...) -> ...;

    fn alloc<T>(&self) -> ... where ... {
        // implementation here
    }

    etc...
}
```

Then code that currently uses `ArenaExt` would only need to use `Arena`.

Reply:
Unfortunately, that would make `Arena` not be object-safe, and I'm not excited about using the `where Sized` trick.

Reply:
SGTM